presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -633,6 +633,13 @@ folly::Future<std::unique_ptr<Result>> TaskManager::getResults(
 
     for (;;) {
       if (prestoTask->taskStarted) {
+        // If the task is failed, then return an error. Before accessing velox
+        // task state, we need to test this as velox task can be null if there
+        // is an error present.
+        if (prestoTask->error != nullptr) {
```

Comment:
There is some race conditions in prestoTask fields. So consider to change 'taskStarted' and 'error' to an atomic which seems to be necessary if we enabling tsan on presto cpp tests which might help us to catch this sort of condition in tests.
```
// NOTE: 'error' in PrestoTask might have been set here by async query failure code path even though we have checked 'error' right above this loop. Also createOrUpdateErrorTask sets both 'error' and 'taskStarted' but leave velox 'task' empty in async query failure case.
```
Also consider to create a unit test using testValue facility provided by Velox.